brave/brave-core
# feat(wallet): transaction simulations

Resolves https://github.com/brave/brave-browser/issues/24269
Resolves https://github.com/brave/brave-browser/issues/24271
Resolves https://github.com/brave/brave-browser/issues/34617
Resolves https://github.com/brave/brave-browser/issues/31949
Resolves https://github.com/brave/brave-browser/issues/32205
Resolves https://github.com/brave/brave-browser/issues/34981
Resolves https://github.com/brave/brave-browser/issues/25755
Resolves https://github.com/brave/brave-browser/issues/26613
Resolves https://github.com/brave/brave-browser/issues/39512

Security/Privacy review: https://github.com/brave/reviews/issues/1551



## Submitter Checklist:

  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`

## Reviewer Checklist:


## After-merge Checklist:

  changes has landed on

## Test Plan:

Note: Brave EVM Brave Swap, EVM CoW Swap, & Matcha (0x) swap transactions will skip simulation since we can determine the changes without an external service.

Scenario 1: Opt-in modal displays and feature activates as expected
1. Enable Transaction simulations flag: chrome://flags/#brave-wallet-enable-transaction-simulations
2. Create a new transaction
3. Test approval/denial of the opt-in modal
4. Test that opt-in modal appears if "Ask" is selected as the option in brave://settings/web3 (See the results of transactions before you sign.)
5. Test that opt-in modal does not appear and simulation is not occurring if "No" is selected as the option in brave://settings/web3 (See the results of transactions before you sign.)
6. Test that opt-in modal does not appear and simulation is occurring if "Yes" is selected as the option in brave://settings/web3 (See the results of transactions before you sign.)

Scenario 2: Simulation occurs as expected
1. Enable transaction simulation as outlined in scenario 1.
2. Queue the following transactions
- Solana Send (NFT)
- Solana Send (SPL token)
- Jupiter swap
- EVM Send (Erc20)
- EVM Send (ERC721)
- EVM Token approval (Erc20)
- EVM token approval (ERC721)
- Uniswap swap
3. open the panel and observe the queued transactions
4.  Simulation should occur and show expected results for each transaction
